lm_diagnoser/interventions/weakly_supervised.py

```diff
@@ -0,0 +1,360 @@
+                                   label: Tensor) -> _Loss:
+        """
+        Define in this function how the loss of the Diagnostic Classifier's prediction w.r.t to the loss is calculated.
+        Should return a subclass of PyTorch's _Loss object like NLLLoss or CrossEntropyLoss.
+
+        Parameters
+        ----------
+        prediction: Tensor
+            Prediction of the diagnostic classifier based on the current time step's activations.
+        label: Tensor
+            Actual label to compare the prediction to.
+        """
+        ...
+
+    @overrides
+    def intervention_func(self,
+                          inp: str,
+                          prev_activations: FullActivationDict,
+                          out: Tensor,
+                          activations: FullActivationDict,
+                          **additional: Dict) -> Tuple[Tensor, FullActivationDict]:
+        """
+        Conduct an intervention based on weak supervision signal.
+
+        Parameters
+        ----------
+        inp: str
+            Current input token.
+        prev_activations: FullActivationDict
+            Activations of the previous time step.
+        out: Tensor
+            Output Tensor of current time step.
+        activations: FullActivationDict
+            Activations of current time step.
+        additional: dict
+            Dictionary of additional information delivered via keyword arguments.
+
+        Returns
+        -------
+        out: Tensor
+            Re-decoded output Tensor of current time step.
+        activations: FullActivationDict
+            Activations of current time step after interventions.
+        """
+
+        dc = self.select_diagnostic_classifier(inp, prev_activations, **additional)
+        weights, bias = self.diagnostic_classifier_to_vars(dc)
+        label = torch.tensor([additional["label"]], dtype=torch.float)
+
+        # Calculate gradient of the diagnostic classifier's prediction w.r.t. the current activations
+        current_activations = activations[self.num_topmost_layer]["hx"]
+        current_activations = self._wrap_in_var(current_activations, requires_grad=True)
+        params = [current_activations]
+        optimizer = SGD(params, lr=self.step_size)
+        optimizer.zero_grad()
+
+        prediction = sigmoid(weights @ current_activations + bias).unsqueeze(0)
+        mask = self.dc_trigger_func(prev_activations, activations, out, prediction, **additional)
+        loss = self.diagnostic_classifier_loss(prediction, label)
+        loss.backward()
+        gradient = current_activations.grad
+
+        # Manual (masked) update step
+        new_activations = current_activations + self.step_size * gradient * mask
+
+        # Repeat decoding step with adjusted activations
+        out: Tensor = self.model.w_decoder @ new_activations + self.model.b_decoder
+
+        return out, activations
+
+
+class LanguageModelInterventionMechanism(WeaklySupervisedInterventionMechanism):
```

Comment:
I would move this (and the other class below) to their own file, as this file is becoming quite lengthy already.

Reply:
Perhaps, but if you think they can be useful for different datasets/experiments I'm fine with keeping them there